### ISSUE TITLE
 Incorrect Stack Trace Formatting in Release Mode on .NET MAUI

### ISSUE THREAD

**Issue by am-kh** [xamarin/xamarin-macios#19417]
**Issue:**
When running my .NET MAUI (iOS) application in Release mode, the generated stack trace is incomplete and lacks crucial details compared to Debug mode. This poses challenges in identifying and resolving issues, particularly related to `System.ArgumentOutOfRangeException`.

**Details:**
- **Framework Used:** Microsoft.iOS (native)
- **.NET Versions Tested:** .NET 7 and .NET 8 RC2
- **iOS Devices Tested:** iPad running iOS 16 and iOS 17
- **IDE:** Visual Studio for Mac version 17.6.6
- **macOS Version:** Sonoma 14.0

**Stack Trace in Debug Mode (Correct):**
```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1[...].get_Item(Int32 index)
   at Confidential.Framework.Business.Common.ObservableCollectionFast`1[...].get_Item(Int32 index) in /Users/confidentialuser/PathToProject/Framework/Business/Confidential.Framework.Business.Common/ObservableCollectionFast.cs:line 38
   at System.Linq.Enumerable.ElementAt[ISaleOrderItemData](IEnumerable`1 source, Int32 index)
   at ConfidentialiPad.Views.Modules.Sales.Views.SalesTableView`1.SalesTableViewSource[...].GetCell(UITableView tableView, NSIndexPath indexPath) in /Users/confidentialuser/PathToProject/ConfidentialMAUI/ConfidentialMAUI.iOS/Views/Modules/Sales/Views/SalesTableView.cs:line 123
   at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName)
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)
   at Program.<Main>$(String[] args) in /Users/confidentialuser/PathToProject/ConfidentialMAUI/ConfidentialMAUI.iOS/Main.cs:line 75
```

**Stack Trace in Release Mode (Incorrect):**
```
System.ArgumentOutOfRangeException: ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index
   at ObjCRuntime.Runtime.ThrowException(IntPtr gchandle)
   at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName)
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)
   at Program.<Main>$(String[] args)
```

**Steps to Reproduce:**
1. Build the .NET MAUI application in Release mode.
2. Encounter an `ArgumentOutOfRangeException` scenario.
3. Observe the incomplete and uninformative stack trace.

**Expected Behavior:**
The stack trace in Release mode should provide detailed information, similar to the one observed in Debug mode, aiding in effective issue resolution.

**Environment:**
- macOS version: Sonoma 14.0
- IDE (Visual Studio for Mac) version: 17.6.6

**Additional Information:**
Include any additional context or information that might be relevant to the issue.

**Project Configuration:**
```xml
<PropertyGroup>
    <TargetFramework>net7.0-ios</TargetFramework>
    <UseMauiEssentials>true</UseMauiEssentials>
    <UseInterpreter>True</UseInterpreter>
    <OutputType>Exe</OutputType>
    <Nullable>enable</Nullable>
    <WarningsAsErrors>Nullable</WarningsAsErrors>
    <ImplicitUsings>true</ImplicitUsings>
    <SupportedOSPlatformVersion>16.0</SupportedOSPlatformVersion>
    <AssemblyVersion>7.0.0.0</AssemblyVersion>
    <FileVersion>7.0.0.0</FileVersion>
    <CodesignProvision></CodesignProvision>
</PropertyGroup>

<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
    <ProvisioningType>automatic</ProvisioningType>
    <CodesignKey>iPhone Developer</CodesignKey>
    <AdditionalArgs></AdditionalArgs>
    <MtouchLink>None</MtouchLink>
    <MtouchUseLlvm>False</MtouchUseLlvm>
    <CreatePackage>false</CreatePackage>
    <CodesignProvision>iOS Team Provisioning Profile: *</CodesignProvision>
</PropertyGroup>

<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <ProvisioningType>manual</ProvisioningType>
    <RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
    <MtouchLink>None</MtouchLink>
    <CreatePackage>false</CreatePackage>
    <CodesignKey>iPhone Distribution: CERT (MY_CERTIFICATE)</CodesignKey>
    <CodesignProvision>Confidential-iPad-7</CodesignProvision>
    <BuildIpa>True</BuildIpa>
</PropertyGroup>
```

Thank you for your help.

**Comment by rolfbjarne** [comment]
I'm not able to reproduce this, the following code:

```cs
	void A ()
	{
		Console.WriteLine ("A");
		B ();
	}
	void B ()
	{
		Console.WriteLine ("B");
		C ();
	}
	void C ()
	{
		Console.WriteLine ("C");
		D ();
	}
	void D ()
	{
		Console.WriteLine ("D");
		E ();
	}
	void E ()
	{
		var list = new List<string> ();
		Console.WriteLine (list [42]);
	}
```

produced:

```
 *** Terminating app due to uncaught exception 'System.ArgumentOutOfRangeException', reason: 'ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index (System.ArgumentOutOfRangeException)
   at ios_plain.AppDelegate.E()
   at ios_plain.AppDelegate.D()
   at ios_plain.AppDelegate.C()
   at ios_plain.AppDelegate.B()
   at ios_plain.AppDelegate.A()
   at ios_plain.AppDelegate.FinishedLaunching(UIApplication application, NSDictionary launchOptions)
```

Please attach a complete test project we can use to reproduce this.

**Comment by am-kh** [comment]
Thank you for your prompt response. While delving deeper into the issue, I've identified a specific scenario that may shed light on the stack trace problem.

It appears that when an exception occurs on the UI thread, the stack trace information is not being captured, resulting in a lack of details. However, if the exception takes place on a thread other than the UI thread, the stack trace is correctly provided.
I've created a minimal project that reproduces the problem. You can find the project [here](https://github.com/am-kh/StackTraceIssue).

```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at UIKit.UIApplication.UIApplicationMain(Int32 , String[] , IntPtr , IntPtr )
   at UIKit.UIApplication.Main(String[] , Type , Type )
   at Program.<Main>$(String[] args) in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/Main.cs:line 12
```


```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Item(Int32 )
   at CrashiOS.RootViewController.GetItem() in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/RootViewController.cs:line 19
   at CrashiOS.RootViewController.<GetItemAsync>b__3_0() in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/RootViewController.cs:line 24
   at System.Threading.Tasks.Task`1[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__273_0(Object )
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread , ExecutionContext , ContextCallback , Object )
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread , ExecutionContext , ContextCallback , Object )
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& , Thread )
--- End of stack trace from previous location ---
   at CrashiOS.RootViewController.GetItemAsync() in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/RootViewController.cs:line 24
   at CrashiOS.RootViewController.<.ctor>b__1_1(Object _, EventArgs _) in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/RootViewController.cs:line 13
   at UIKit.UIApplication.UIApplicationMain(Int32 , String[] , IntPtr , IntPtr )
   at UIKit.UIApplication.Main(String[] , Type , Type )
   at Program.<Main>$(String[] args) in /Users/aminekhiyat/Projects/CrashiOS/CrashiOS/Main.cs:line 12
```

**Comment by rolfbjarne** [comment]
OK, I can reproduce now, this happens because of how we marshal exceptions between Objective-C and managed code.

Basically we convert managed exceptions into Objective-C exceptions when going through native frames, and vice versa the other way, and after a few of these transitions the original stack trace is lost.

As a workaround, you can be notified when exception marshalling occurs, and collect those stack traces.

Example:

```cs
using ObjCRuntime;

Runtime.MarshalManagedException += (object sender, MarshalManagedExceptionEventArgs args) =>
{
    Console.WriteLine ("Marshaling managed exception");
    Console.WriteLine ("    Exception: {0}", args.Exception);
    Console.WriteLine ("    Mode: {0}", args.ExceptionMode);
    Thread.Sleep (5000);
};

Runtime.MarshalObjectiveCException += (object sender, MarshalObjectiveCExceptionEventArgs args) =>
{
    Console.WriteLine ("Marshaling Objective-C exception");
    Console.WriteLine ("    Exception: {0}", args.Exception);
    Console.WriteLine ("    Mode: {0}", args.ExceptionMode);
};
```

And this will show something like this for your example project:

```
Marshaling managed exception
    Exception: System.ArgumentOutOfRangeException: ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index
    at System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Item(Int32 index)
    at CrashiOS.RootViewController.GetItem()
    at CrashiOS.RootViewController.<.ctor>b__1_0(Object _, EventArgs _)
    at UIKit.UIBarButtonItem.Callback.Call(NSObject sender)
```

**Comment by am-kh** [comment]
Thank you for the explanations. I tried to implement your suggestions, but it's not working. The `MarshalManagedException` and `MarshalObjectiveCException` events are never raised.

```csharp
using System;
using System.Diagnostics;
using System.Threading;
using CrashiOS;
using ObjCRuntime;
using Serilog;

class Program
{
    static void Main(string[] args)
    {
        Log.Logger = new LoggerConfiguration()
            .WriteTo.Console()
            .WriteTo.Debug()
            .MinimumLevel.Information()
            .CreateLogger();

        Runtime.MarshalManagedException += (object sender, MarshalManagedExceptionEventArgs args) =>
        {
            Console.WriteLine("Marshaling managed exception");
            Console.WriteLine($"    Exception: {args.Exception}");
            Console.WriteLine($"    Mode: {args.ExceptionMode}");
            Thread.Sleep(5000);
        };

        Runtime.MarshalObjectiveCException += (object sender, MarshalObjectiveCExceptionEventArgs args) =>
        {
            Console.WriteLine("Marshaling Objective-C exception");
            Console.WriteLine($"    Exception: {args.Exception}");
            Console.WriteLine($"    Mode: {args.ExceptionMode}");
        };

        AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;

        UIApplication.Main(args, null, typeof(AppDelegate));
    }

    static void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
    {
        var ex = (Exception)e.ExceptionObject;
        Log.Fatal(ex, ex.Message);
        Debugger.Break();
    }
}
```

**Comment by rolfbjarne** [comment]
This diff worked for me in the simulator: https://gist.github.com/rolfbjarne/8dba7cf10e13bff8bd693b3e2c80e04c

**Comment by marcoburato-ecutek** [comment]
I believe I just hit this issue. I ported our app to .NET 8 and a NRE thrown in the main thread shows this stack trace:

    System.NullReferenceException: Object reference not set to an instance of an object
    at ObjCRuntime.Runtime.ThrowException(IntPtr gchandle)
    at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName)
    at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)


This kind of crash doesn't even show up in the AppCenter Diagnostics.
The TestFlight crash report shows no managed stack at all:

    ...

    Exception Type:  EXC_BAD_ACCESS (SIGABRT)
    Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000000000000
    Exception Codes: 0x0000000000000001, 0x0000000000000000
    VM Region Info: 0 is not in any region.  Bytes before following region: 4300242944
        REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
        UNUSED SPACE AT START
    --->  
        __TEXT                   100508000-10277c000 [ 34.5M] r-x/r-x SM=COW  <redacted>
    Triggered by Thread:  0


    Kernel Triage:
    VM - (arg = 0x3) mach_vm_allocate_kernel failed within call to vm_map_enter


    Thread 0 name:
    Thread 0 name:
    Thread 0 Crashed:
    0   libsystem_kernel.dylib        	0x00000001d95dd01c __pthread_kill + 8 (:-1)
    1   libsystem_pthread.dylib       	0x00000001fbb29680 pthread_kill + 268 (pthread.c:1681)
    2   libsystem_c.dylib             	0x000000019a13db90 abort + 180 (abort.c:118)
    3   <redacted>                    	0x0000000102453284 0x100508000 + 32813700
    4   <redacted>                    	0x000000010260249c 0x100508000 + 34579612
    5   <redacted>                    	0x000000010269cb84 0x100508000 + 35212164
    6   <redacted>                    	0x000000010245b85c 0x100508000 + 32847964
    7   <redacted>                    	0x000000010277e1fc 0x100508000 + 36135420
    8   dyld                          	0x00000001b4b4edcc start + 2240 (dyldMain.cpp:1269)

    ...


I see the fix for this issue is planned for .NET 9. Surely this is a major bug that needs to be fixed in the next .NET 8 maintenance release?

**Comment by marcoburato-ecutek** [comment]
@rolfbjarne Any idea when this will be fixed? It's stalling my migration from Xamarin to .NET.

**Comment by rolfbjarne** [comment]
@marcoburato-ecutek no idea yet.

Adding event handlers for when exceptions are marshalled doesn't help (https://github.com/xamarin/xamarin-macios/issues/19417#issuecomment-1807927025)?

**Comment by marcoburato-ecutek** [comment]
@rolfbjarne My understanding is that those event handlers would let me write stack traces somewhere (ie. console, log file) but the stack strace in the exception would still be incorrect, which is what gets used by the crash reporting library.

That would make crash reports all contain the same dummy trace, which would be useless. And unrelated reports would likely get grouped together too (ie. all NREs, since the exception description and trace would be the same).

**Comment by rolfbjarne** [comment]
> @rolfbjarne My understanding is that those event handlers would let me write stack traces somewhere (ie. console, log file) but the stack strace in the exception would still be incorrect, which is what gets used by the crash reporting library.
> 
> That would make crash reports all contain the same dummy trace, which would be useless. And unrelated reports would likely get grouped together too (ie. all NREs, since the exception description and trace would be the same).

Yes, that's a good point.

I'll see what I can do.

**Comment by dotMorten** [comment]
I'm more concerned the error messages change in release mode to near-gibberish for lots of different errors:
Debug: `System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')`
Release: `System.ArgumentOutOfRangeException: ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index`

The immediate effect is this makes unit testing exception messages harder, but we also do know error message does often end up in user UI (whether that's good or bad) and automated error reporting, so it would be nice if the messages still makes sense. Would this PR also address that?

**Comment by rolfbjarne** [comment]
> I'm more concerned the error messages change in release mode to near-gibberish for lots of different errors: Debug: `System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')` Release: `System.ArgumentOutOfRangeException: ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index`
> 
> The immediate effect is this makes unit testing exception messages harder, but we also do know error message often end up in user UI (whether that's good or bad) and automated error reporting, so it would be nice if the messages still makes sense. Would this PR also address that?

That's a feature to minimize app size.

You can turn it off by doing this:

```xml
<PropertyGroup>
    <UseSystemResourceKeys>false</UseSystemResourceKeys>
</PropertyGroup>
```

Ref: https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-8-0#trimming-framework-library-features

**Comment by jstrawther** [comment]
@rolfbjarne Thank you for the timely fix.  Will this be included in an 8.x release?

**Comment by rolfbjarne** [comment]
> @rolfbjarne Thank you for the timely fix. Will this be included in an 8.x release?

Yes, we'll include this fix in an upcoming service release (not exactly sure when that will be at the moment).

**Comment by malsabi** [comment]
@rolfbjarne  So I use the following snippet:

```
<PropertyGroup>
    <UseSystemResourceKeys>false</UseSystemResourceKeys>
</PropertyGroup>
```

to avoid this exception?
```Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')```
I'm using .NET MAUI app and I currently have this exception thrown when using CollectionView on iOS.

It's really fruastrating if there's any workaround please let me know.

**Comment by rolfbjarne** [comment]
> @rolfbjarne So I use the following snippet:
> 
> ```
> <PropertyGroup>
>     <UseSystemResourceKeys>false</UseSystemResourceKeys>
> </PropertyGroup>
> ```
> 
> to avoid this exception? `Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')` I'm using .NET MAUI app and I currently have this exception thrown when using CollectionView on iOS.
> 
> It's really fruastrating if there's any workaround please let me know.

No, `UseSystemResourceKeys=false` only changes the exception message (from "ArgumentOutOfRange_IndexMustBeLess Arg_ParamName_Name, index" to "Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')"), it does not affect whether the exception is thrown or not.

In order to fix the exception you need to figure out why the exception is thrown.

If you think there's a bug in MAUI causing the exception, you can create a test project and file an issue with the test project at https://github.com/dotnet/maui/issues/new/choose, and the corresponding team will have a look.

**Comment by jurganson** [comment]
@rolfbjarne Do you know which release this fix will be part of/how I can get a hold of this fix? I can see some things merged to service+backported but the milestone of this issue is set to .NET 9, and I just want to get my hands on this fix ASAP 😛

**Comment by jurganson** [comment]
@rolfbjarne Thanks for the answer. Do you know if the dependencies of Microsoft.SDK.iOS update of .NET MAUI is regular? It seems like currently on [.NET MAUI master](https://github.com/dotnet/maui/blob/1c8e512882fd5f5d8b078a6a407bfefbcac957b7/eng/Versions.props#L56) that its pointing at 17.2.8022, which is older than [17.2.8043](https://www.nuget.org/packages/Microsoft.iOS.Sdk/17.2.8043) which came out back in March 12, even though there was a service release of .NET MAUI last week, 8.0.40 SR5.

**Comment by rolfbjarne** [comment]
> @rolfbjarne Thanks for the answer. Do you know if the dependencies of Microsoft.SDK.iOS update of .NET MAUI is regular? It seems like currently on [.NET MAUI master](https://github.com/dotnet/maui/blob/1c8e512882fd5f5d8b078a6a407bfefbcac957b7/eng/Versions.props#L56) that its pointing at 17.2.8022, which is older than [17.2.8043](https://www.nuget.org/packages/Microsoft.iOS.Sdk/17.2.8043) which came out back in March 12, even though there was a service release of .NET MAUI last week, 8.0.40 SR5.

Generally we release service updates at the same time as MAUI does, but as you've seen it doesn't always happen.

In any case, which version of the iOS SDK MAUI points to is somewhat irrelevant, as you'll always get the latest version we release.

**Comment by jurganson** [comment]
> In any case, which version of the iOS SDK MAUI points to is somewhat irrelevant, as you'll always get the latest version we release.

@rolfbjarne Alright, perhaps then I simply misunderstand how the updates are deployed.

Right now, I am running Visual Studio 17.9.6 (Build number 17.9.34728.123).

I am currently running a ported Xamarin multi-project in MAUI. I have not explicitly installed any Microsoft.iOS.Sdk, rather simply the one that comes with my Visual Studio MAUI workload. However, if I understand how to see the version correctly, what I am seeing is 17.2.8004, which was released back in January 17th.

```
Installed Workload Id      Manifest Version       Installation Source
---------------------------------------------------------------------
ios                        17.2.8004/8.0.100      VS 17.9.34728.123
```

My current MAUI version is set through my project files using `<MauiVersion>`, which is currently set to 8.0.21.

I was under the impression that the iOS workload would be updated whenever I updated .NET MAUI as it's a dependency. However, I see the installation source is Visual Studio. Although the version I am using (17.9.6) is from April 9, 2024, which in theory means it could contain Microsoft.iOS.Sdk 8043, or at least 8022.

So who/what controls the Microsoft.iOS.Sdk version and how do I update it?
1. Is it controlled by Visual Studio, and thus I need to update Visual Studio whenever I want a newer iOS workload?
2. Is it controlled by .NET MAUI, which I use the iOS workload through?
3. Do I control it myself and can simply update it regardless of .NET MAUI or Visual Studio?

Sorry for the long line of questioning, but I really want to nail down how the different versions work for me and my team 😄
